frontend/src/metabase-lib/parameters/utils/click-behavior.ts

```diff
@@ -0,0 +1,397 @@
+  })[];
+}): ValueAndColumnForColumnNameDate {
+  const column = [
+    ...dimensions,
+    ...data.map(d => ({
+      column: d.col,
+      // When the data is changed to a display value for use in tooltips, we can set clickBehaviorValue to the raw value for filtering.
+      value: d.clickBehaviorValue || d.value,
+    })),
+  ]
+    .filter(d => d.column != null)
+    .reduce<ValueAndColumnForColumnNameDate["column"]>(
+      (acc, { column, value }) => {
+        if (!column) {
+          return acc;
+        }
+
+        const name = column.name.toLowerCase();
+
+        if (acc[name] === undefined) {
+          return { ...acc, [name]: { value, column } };
+        }
+
+        return acc;
+      },
+      {},
+    );
+
+  const dashboardParameters = (dashboard?.parameters || []).filter(
+    ({ slug }) => parameterValuesBySlug[slug] != null,
+  );
+
+  const parameterByName = Object.fromEntries(
+    dashboardParameters.map(({ name, slug }) => [
+      name.toLowerCase(),
+      { value: parameterValuesBySlug[slug] },
+    ]),
+  );
+
+  const parameterBySlug = _.mapObject(parameterValuesBySlug, value => ({
+    value,
+  }));
+
+  const parameter = Object.fromEntries(
+    dashboardParameters.map(({ id, slug }) => [
+      id,
+      { value: parameterValuesBySlug[slug] },
+    ]),
+  );
+
+  const userAttribute = Object.fromEntries(
+    (userAttributes || []).map(value => [value, { value }]),
+  );
+
+  return { column, parameter, parameterByName, parameterBySlug, userAttribute };
+}
+
+const { Text, Number, Temporal } = TYPE;
+
+function notRelativeDateOrRange({ type }: Parameter) {
+  return type !== "date/range" && type !== "date/relative";
+}
+
+export function getTargetsForQuestion(question: Question): Target[] {
+  const query = question.query();
+  return [...query.dimensionOptions().all(), ...query.variables()].map(o => {
+    let id, target: ClickBehaviorTarget;
+    if (o instanceof TemplateTagVariable || o instanceof TemplateTagDimension) {
+      let name;
+      ({ id, name } = o.tag());
+      target = { type: "variable", id: name };
+    } else if ("mbql" in o) {
+      const dimension: ClickBehaviorDimensionTarget["dimension"] = [
+        "dimension",
+        o.mbql(),
+      ];
+      id = JSON.stringify(dimension);
+      target = { type: "dimension", id, dimension };
+    } else {
+      throw new Error("Unknown target type");
+    }
+    let parentType: string | undefined | null;
+    let parameterSourceFilter: SourceFilters["parameter"] = () => true;
+    if (o instanceof TemplateTagVariable) {
+      const type = o.tag()?.type;
+      parentType = type
+        ? {
+            card: undefined,
+            dimension: undefined,
+            snippet: undefined,
+            text: Text,
+            number: Number,
+            date: Temporal,
+          }[type]
+        : undefined;
+      parameterSourceFilter = parameter =>
+        variableFilterForParameter(parameter)(o);
+    } else if ("field" in o) {
```

Comment:
Why do you change the condition from `o.field() != null` to `"field" in o`. I've tried reverting this back and TypeScript doesn't complain anything.

Reply:
5c77a4d7fcb3bcf44f66bfda4502f589ca9553b6